### PR TITLE
Add float nan box and tests

### LINKED ISSUE #97: Riscv64-simulator can't handle a  float value which isn't  NaN-boxed.
**Describe the bug**
On riscv64, a float value is stored into 64-bit f-register. It should be NaN boxed, such as `zero` boxed to `0xFFFFFFFF00000000ULL`. If not boxed, it will be treated as a NaN value on hifive-unleashed. So our simulator should also treated  it as NaN.
refer https://groups.google.com/a/groups.riscv.org/g/sw-dev/c/mfGw77uQYwc/m/HhOb3QUvBAAJ
 
**To Reproduce**
relating #74 

**Expected behavior**
None

**Errors/Logs**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by LOGQTAINIA**
Fix #97 and #102

### CODE REVIEW

**Comment by penguinwu** [comment]
There are a quite a few cctest test failures, please address those failures first

**Comment by LOGQTAINIA** [comment]
I found the error logs are the same as #113 . 

Maybe the change of simulator is no problem.

**Comment by LOGQTAINIA** [comment]
**Errors/Logs**

```
[11:08|%  79|+ 5219|-   1]: cctest/test-run-wasm-simd/RunWasm_SimdF32x4AddW...
                                                                              
=== cctest/test-run-wasm-simd/RunWasm_SimdF32x4AddWithI32x4_simd_lowered ===
#
# Fatal error in ../../test/cctest/wasm/test-run-wasm-simd.cc, line 3064
# Check failed: 1 == r.Call() (1 vs. 0).
#
#
#
#FailureMessage Object: 0x7ffc2d573da0
==== C stack trace ===============================

    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d4abd3) [0x559c8fed2bd3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d49d4d) [0x559c8fed1d4d]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3ab98b3) [0x559c8fc418b3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x18a0216) [0x559c8da28216]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x18a0064) [0x559c8da28064]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x11ff199) [0x559c8d387199]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x12005dc) [0x559c8d3885dc]
    /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f34ee84bb97]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(_start+0x2a) [0x559c8d386f0a]
Aborted (core dumped)
Command: out/riscv64.sim/cctest test-run-wasm-simd/RunWasm_SimdF32x4AddWithI32x4_simd_lowered --random-seed=167668032 --nohard-abort --enable-slow-asserts --verify-heap --testing-d8-test-runner


[11:08|%  79|+ 5219|-   2]: cctest/test-run-wasm-simd/RunWasm_SimdF32x4AddW...
                                                                              
=== cctest/test-run-wasm-simd/RunWasm_SimdF32x4AddWithI32x4_turbofan ===
#
# Fatal error in ../../test/cctest/wasm/test-run-wasm-simd.cc, line 3064
# Check failed: 1 == r.Call() (1 vs. 0).
#
#
#
#FailureMessage Object: 0x7ffc95a30200
==== C stack trace ===============================

    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d4abd3) [0x555ad607fbd3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d49d4d) [0x555ad607ed4d]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3ab98b3) [0x555ad5dee8b3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x18a0216) [0x555ad3bd5216]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x18a0001) [0x555ad3bd5001]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x11ff199) [0x555ad3534199]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x12005dc) [0x555ad35355dc]
    /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f103be98b97]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(_start+0x2a) [0x555ad3533f0a]
Aborted (core dumped)
Command: out/riscv64.sim/cctest test-run-wasm-simd/RunWasm_SimdF32x4AddWithI32x4_turbofan --random-seed=167668032 --nohard-abort --enable-slow-asserts --verify-heap --testing-d8-test-runner


[11:08|%  79|+ 5223|-   3]: cctest/test-run-wasm-simd/RunWasm_SimdF32x4Extr...
                                                                              
=== cctest/test-run-wasm-simd/RunWasm_SimdF32x4ExtractWithI32x4_simd_lowered ===
#
# Fatal error in ../../test/cctest/wasm/test-run-wasm-simd.cc, line 3033
# Check failed: 1 == r.Call() (1 vs. 0).
#
#
#
#FailureMessage Object: 0x7ffc6d451130
==== C stack trace ===============================

    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d4abd3) [0x560ae3c31bd3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d49d4d) [0x560ae3c30d4d]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3ab98b3) [0x560ae39a08b3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x189faf1) [0x560ae1786af1]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x189f954) [0x560ae1786954]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x11ff199) [0x560ae10e6199]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x12005dc) [0x560ae10e75dc]
    /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7fd870359b97]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(_start+0x2a) [0x560ae10e5f0a]
Aborted (core dumped)
Command: out/riscv64.sim/cctest test-run-wasm-simd/RunWasm_SimdF32x4ExtractWithI32x4_simd_lowered --random-seed=167668032 --nohard-abort --enable-slow-asserts --verify-heap --testing-d8-test-runner

                                                                              
[11:08|%  79|+ 5223|-   4]: cctest/test-run-wasm-simd/RunWasm_SimdF32x4Extr...

=== cctest/test-run-wasm-simd/RunWasm_SimdF32x4ExtractWithI32x4_turbofan ===
#
# Fatal error in ../../test/cctest/wasm/test-run-wasm-simd.cc, line 3033
# Check failed: 1 == r.Call() (1 vs. 0).
#
#
#
#FailureMessage Object: 0x7ffc24698530
==== C stack trace ===============================

    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d4abd3) [0x561f5b3f5bd3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d49d4d) [0x561f5b3f4d4d]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3ab98b3) [0x561f5b1648b3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x189faf1) [0x561f58f4aaf1]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x189f8f1) [0x561f58f4a8f1]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x11ff199) [0x561f588aa199]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x12005dc) [0x561f588ab5dc]
    /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7fdc39611b97]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(_start+0x2a) [0x561f588a9f0a]
Aborted (core dumped)
Command: out/riscv64.sim/cctest test-run-wasm-simd/RunWasm_SimdF32x4ExtractWithI32x4_turbofan --random-seed=167668032 --nohard-abort --enable-slow-asserts --verify-heap --testing-d8-test-runner


[11:09|%  79|+ 5227|-   5]: cctest/test-run-wasm-simd/RunWasm_SimdF32x4GetG...
                                                                              
=== cctest/test-run-wasm-simd/RunWasm_SimdF32x4GetGlobal_simd_lowered ===
#
# Fatal error in ../../test/cctest/wasm/test-run-wasm-simd.cc, line 3247
# Check failed: 1 == r.Call(0) (1 vs. 0).
#
#
#
#FailureMessage Object: 0x7ffedb15fbb0
==== C stack trace ===============================

    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d4abd3) [0x55e79fac6bd3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d49d4d) [0x55e79fac5d4d]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3ab98b3) [0x55e79f8358b3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x18a18f9) [0x55e79d61d8f9]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x18a1664) [0x55e79d61d664]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x11ff199) [0x55e79cf7b199]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x12005dc) [0x55e79cf7c5dc]
    /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7fee369e1b97]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(_start+0x2a) [0x55e79cf7af0a]
Aborted (core dumped)
Command: out/riscv64.sim/cctest test-run-wasm-simd/RunWasm_SimdF32x4GetGlobal_simd_lowered --random-seed=167668032 --nohard-abort --enable-slow-asserts --verify-heap --testing-d8-test-runner


[11:09|%  79|+ 5227|-   6]: cctest/test-run-wasm-simd/RunWasm_SimdF32x4GetG...
                                                                              
=== cctest/test-run-wasm-simd/RunWasm_SimdF32x4GetGlobal_turbofan ===
#
# Fatal error in ../../test/cctest/wasm/test-run-wasm-simd.cc, line 3247
# Check failed: 1 == r.Call(0) (1 vs. 0).
#
#
#
#FailureMessage Object: 0x7ffcb1543390
==== C stack trace ===============================

    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d4abd3) [0x5567006b4bd3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3d49d4d) [0x5567006b3d4d]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x3ab98b3) [0x5567004238b3]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x18a18f9) [0x5566fe20b8f9]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x18a1601) [0x5566fe20b601]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x11ff199) [0x5566fdb69199]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(+0x12005dc) [0x5566fdb6a5dc]
    /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f989a519b97]
    /home/runner/work/v8/v8/v8/out/riscv64.sim/cctest(_start+0x2a) [0x5566fdb68f0a]
Aborted (core dumped)
Command: out/riscv64.sim/cctest test-run-wasm-simd/RunWasm_SimdF32x4GetGlobal_turbofan --random-seed=167668032 --nohard-abort --enable-slow-asserts --verify-heap --testing-d8-test-runner

```

File: src/execution/riscv64/simulator-riscv64.cc [hunk 1]
```diff
@@ -914,7 +914,7 @@ void Simulator::set_fpu_register_hi_word(int fpureg, int32_t value) {
 
 void Simulator::set_fpu_register_float(int fpureg, float value) {
   DCHECK((fpureg >= 0) && (fpureg < kNumFPURegisters));
-  *bit_cast<float*>(&FPUregisters_[fpureg]) = value;
+  FPUregisters_[fpureg] = boxF32(value);
```
Comment (penguinwu) [inline src/execution/riscv64/simulator-riscv64.cc:917]: To make the naming style consistent, suggest `boxF32()` -> `box_float()`

File: src/execution/riscv64/simulator-riscv64.cc [hunk 2]
```diff
@@ -967,6 +967,9 @@ int32_t Simulator::get_fpu_register_hi_word(int fpureg) const {
 
 float Simulator::get_fpu_register_float(int fpureg) const {
   DCHECK((fpureg >= 0) && (fpureg < kNumFPURegisters));
+  if (!isBoxedF32(FPUregisters_[fpureg])) {
```
Comment (penguinwu) [inline src/execution/riscv64/simulator-riscv64.cc:970]: Consistent naming style, `isBoxedF32` -> `is_boxed_float`

File: src/execution/riscv64/simulator-riscv64.cc [hunk 3]
```diff
@@ -2391,7 +2394,7 @@ void Simulator::DecodeRVRFPType() {
         case 0b000: {
           if (instr_.Rs2Value() == 0b00000) {
             // RO_FMV_X_W
-            set_rd(bit_cast<int32_t>(frs1()));
+            set_rd(get_fpu_register_word(rs1_reg()));
```
Comment (penguinwu) [inline src/execution/riscv64/simulator-riscv64.cc:2397]: Though sign-extension is implicitly done in the original implementation. I think adding `sext_xlen` would make the sign-extension explicit (`set_rd(sext_xlen(get_fpu_register_word(rs1_reg()))`) and the code easier to understand.

File: src/execution/riscv64/simulator-riscv64.cc [hunk 4]
```diff
@@ -3035,7 +3038,9 @@ void Simulator::DecodeRVSType() {
 #endif /*V8_TARGET_ARCH_64_BIT*/
     // TODO: use F Extension macro block
     case RO_FSW: {
-      WriteMem<float>(rs1() + s_imm12(), frs2(), instr_.instr());
+      WriteMem<uint32_t>(rs1() + s_imm12(), 
+                         (uint32_t)get_fpu_register_word(rs2_reg()), 
+                         instr_.instr());
```
Comment (penguinwu) [inline src/execution/riscv64/simulator-riscv64.cc:3043]: This is confusing. Why not implement FSW as a floating-point write? I think it can be done as
```
WriteMem<float>(sr1() + s_imm12(),  get_fpu_register_float(rs2_reg()), instr_.instr());
```

File: src/execution/riscv64/simulator-riscv64.h [hunk 5]
```diff
@@ -125,11 +125,15 @@ inline int64_t mulhsu(int64_t a, uint64_t b) {
...
 
 // Floating point helpers
+#define isBoxedF32(v) ((uint32_t)((v >> 32) + 1) == 0)
 #define F32_SIGN ((uint32_t)1 << 31)
 union u32_f32 {
   uint32_t u;
   float f;
 };
+inline int64_t boxF32(float v) { 
+  return ((uint64_t)-1 << 32) | bit_cast<int32_t>(v);
```
Comment (penguinwu) [inline src/execution/riscv64/simulator-riscv64.h:135]: The operator precedence of cast and `<<` can be a bit confusing. `return (0xFFFF_FFFF_0000_000 | bit_cast<int32_t>(v))` would be more straightforward to understand

File: src/execution/riscv64/simulator-riscv64.h [hunk 6]
```diff
@@ -125,11 +125,15 @@ inline int64_t mulhsu(int64_t a, uint64_t b) {
 }
 
 // Floating point helpers
+#define isBoxedF32(v) ((uint32_t)((v >> 32) + 1) == 0)
```
Comment (penguinwu) [inline src/execution/riscv64/simulator-riscv64.h:128]: to be consistent w/ `boxF32`, make `isBoxedF32` a static inline function as well.

File: src/execution/riscv64/simulator-riscv64.cc (hunk 4 above)
Comment (penguinwu) [reply src/execution/riscv64/simulator-riscv64.cc:3043]: The handling of `FLW` needs to nan-box the loaded value.

> Floating-point n-bit transfer operations move external values held in IEEE standard formats into
and out of the f registers, and **comprise floating-point loads and stores (FLn/FSn)** and floatingpoint
move instructions (FMV.n.X/FMV.X.n). **A narrower n-bit transfer, n < FLEN, into the f
registers will create a valid NaN-boxed value**. A narrower n-bit transfer out of the floating-point
registers will transfer the lower n bits of the register ignoring the upper FLEN−n bits.

```
case RO_FLW: {
      int64_t addr = rs1() + imm12();
      float val = ReadMem<float>(addr, instr_.instr());
      set_frd(val, false); <== need to be nan-boxed
      TraceMemRd(addr, val, get_fpu_register(frd_reg()));
      break;
    }
```

File: test/cctest/test-assembler-riscv64.cc [hunk 7]
```diff
@@ -1431,6 +1431,97 @@ TEST(RISCV9) {
...
+  t.r3 = -1;
+  t.r4 = -1;
+  t.r5 = -1;
+  t.fr1 = -1.0f;
+  t.fr2 = -1.0f;
+  f.Call(&t, 0, 0, 0, 0);
+
+  CHECK_EQ(neg_inf, t.a);
+  CHECK_EQ(0.0f, t.b);
+  CHECK_EQ(((uint64_t)-1 << 32) | bit_cast<int32_t>(neg_inf), t.r0);
```
Comment (penguinwu) [inline test/cctest/test-assembler-riscv64.cc:1515]: introduce a helper function 'is_boxed_float`, so that the meaning of the checking is explicit.

File: test/cctest/test-assembler-riscv64.cc [hunk 8]
```diff
@@ -1431,6 +1431,97 @@ TEST(RISCV9) {
...
+  auto f = GeneratedCode<F3>::FromCode(*code);
+  t.a = neg_inf;
+  t.b = 1.234f;
+  t.r1 = -1;
+  t.r2 = -1;
+  t.r3 = -1;
+  t.r4 = -1;
+  t.r5 = -1;
+  t.fr1 = -1.0f;
+  t.fr2 = -1.0f;
```
Comment (penguinwu) [inline test/cctest/test-assembler-riscv64.cc:1510]: This test case is not easy to understand. There is a lot of stuff that is not related to Nan-boxing testing, perhaps things copied from other tests? Could you make it simpler and remove the unrelated tests?

Make sure you also test the nan-boxing of `flw` instructions.

**Comment by penguinwu** [comment]
It is good that we can trigger the error in simulated build. I suspect that,t in v8-riscv, SIMD opcodes are implemented as scalar operations (see `simd-scalar-lowering.cc` that handles `F32x4Add` (which corresponds to `kExprF32x4Add` that is used in the test case. 

Luckily this test runs only 129 instructions under simulator build (use  `--trace-sim` to dump the execution trace), so it would be relatively easy to debug in the simulated build.  Since the test case passes before the nan-boxing fix, you may want to run both versions under --trace-sim and figure out where the wrong result is triggered.

File: src/execution/riscv64/simulator-riscv64.cc (hunk 4 above)
Comment (LOGQTAINIA) [reply src/execution/riscv64/simulator-riscv64.cc:3043]: Before I explain the questions about FLW and FSW, I'd like to explain some things related.

`box float` and `check boxed float` will be done in `set_fpu_register_float` and `get_fpu_register_float`.

```
void Simulator::set_fpu_register_float(int fpureg, float value) {
  DCHECK((fpureg >= 0) && (fpureg < kNumFPURegisters));
  FPUregisters_[fpureg] = box_float(value);
}

float Simulator::get_fpu_register_float(int fpureg) const {
  DCHECK((fpureg >= 0) && (fpureg < kNumFPURegisters));
  if (!is_boxed_float(FPUregisters_[fpureg])) {
    return std::numeric_limits<float>::quiet_NaN();
  }
  return *bit_cast<float*>(const_cast<int64_t*>(&FPUregisters_[fpureg]));
}
```

Back to the questions.

The original implementation of FSW with `frs2` which invokes `get_fpu_register_float`. But the handling of `FSW` should not `check boxed float`.

> A narrower n-bit transfer out of the floating-point
registers will transfer the lower n bits of the register ignoring the upper FLEN−n bits.

As for `FLW`, it's no problem. Because `set_frd` would invoke `set_fpu_register_float` which boxes float number.

```
inline void set_frd(float value, bool trace = true) {
    set_fpu_register_float(rd_reg(), value);
    if (trace) TraceRegWr(get_fpu_register_word(rd_reg()), FLOAT);
  }
```
